File: 0841-19-2685-count-the-number-of-complete-components/0841-19-2685-count-the-number-of-complete-components.py
class Solution:
    def countCompleteComponents(self, n: int, edges: list[list[int]]) -> int:
        vis = [False] * n
        adj = [[] for _ in range(n)]
        
        for edge in edges:
            adj[edge[0]].append(edge[1])
            adj[edge[1]].append(edge[0])
        
        def dfs(u):
            vis[u] = True
            ret = 1
            nonlocal sz
            sz &= len(adj[u])
            for ch in adj[u]:
                if not vis[ch]:
                    ret += dfs(ch)
            return ret
        
        cnt = 0
        for i in range(n):
            if vis[i]:
                continue
            sz = len(adj[i])
            cnt += (dfs(i) == sz + 1)
        
        return cnt

File: 0841-19-2685-count-the-number-of-complete-components/0841-19-2685-count-the-number-of-complete-components.java
class Solution {
    private boolean[] vis;
    private List<Integer>[] adj;
    private int sz;
    
    public int countCompleteComponents(int n, int[][] edges) {
        vis = new boolean[n];
        adj = new ArrayList[n];
        
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        
        for (int[] edge : edges) {
            adj[edge[0]].add(edge[1]);
            adj[edge[1]].add(edge[0]);
        }
        
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            if (vis[i]) continue;
            sz = adj[i].size();
            cnt += (dfs(i) == sz + 1) ? 1 : 0;
        }
        
        return cnt;
    }
    
    private int dfs(int u) {
        vis[u] = true;
        int ret = 1;
        sz &= adj[u].size();
        
        for (int ch : adj[u]) {
            if (!vis[ch]) {
                ret += dfs(ch);
            }
        }
        
        return ret;
    }
}